--- pkgs/libstr-0.0.2/src/str.c	2013-08-06 21:56:46.000000000 -0400
+++ src/str.c	2013-08-07 00:29:59.415271899 -0400
@@ -34,10 +34,14 @@
 
 	if(index < 0 || index > strlen(haystack)) return NULL;
 
-	strncpy(buff, haystack, index);
-	strcpy(buff+index, needle);
-	strcpy(buff+index+strlen(needle), haystack+index);
+	//use our own temp buffer to perform the operation on incase buff == haystack
+	char tmpbuff[strlen(haystack)+strlen(needle)+1];
 
+	strncpy(tmpbuff, haystack, index);
+	strcpy(tmpbuff+index, needle);
+	strcpy(tmpbuff+index+strlen(needle), haystack+index);
+
+	strcpy(buff,tmpbuff);
 	return buff;
 }
 
