--- src/str.c	2013-09-16 14:38:49.147925753 -0400
+++ src/str.c.new	2013-09-16 14:32:25.514582126 -0400
@@ -59,8 +63,17 @@
 char* strsub(char* expr, size_t strt, size_t end, char* buff)
 {
 	if(!expr || !buff) return NULL;
-
-	strncpy(buff,expr+strt,end-strt+1);
-	*(buff+(end-strt+1)) = '\0';
+	
+	if(strt < 0 || strt > end || strt > strlen(expr)-1) return NULL;
+	if(end < 0|| end < strt || end > strlen(expr)-1) return NULL;
+
+	//use a temp buff to allow buff == expr
+	char tmpbuff[end-strt+2];
+
+	//extract the chars from expr
+	strncpy(tmpbuff,expr+strt,end-strt+1);
+	tmpbuff[end-strt+1] = '\0';
+	
+	strncpy(buff,tmpbuff,end-strt+2);
 	return buff;
 }
