--- pkgs/libstr-0.0.2/src/str.c	2013-08-06 21:56:46.000000000 -0400
+++ src/str.c	2013-08-13 11:12:26.165656078 -0400
@@ -34,10 +34,14 @@
 
 	if(index < 0 || index > strlen(haystack)) return NULL;
 
-	strncpy(buff, haystack, index);
-	strcpy(buff+index, needle);
-	strcpy(buff+index+strlen(needle), haystack+index);
+	//use our own temp buffer to perform the operation on incase buff == haystack
+	char tmpbuff[strlen(haystack)+strlen(needle)+1];
 
+	strncpy(tmpbuff, haystack, index);
+	strcpy(tmpbuff+index, needle);
+	strcpy(tmpbuff+index+strlen(needle), haystack+index);
+
+	strcpy(buff,tmpbuff);
 	return buff;
 }
 
@@ -59,8 +63,16 @@
 char* strsub(char* expr, size_t strt, size_t end, char* buff)
 {
 	if(!expr || !buff) return NULL;
-
-	strncpy(buff,expr+strt,end-strt+1);
-	*(buff+(end-strt+1)) = '\0';
+	
+	if(strt < 0 || strt > end || strt > strlen(expr)-1) return NULL;
+	if(end < 0|| end < strt || end > strlen(expr)-1) return NULL;
+
+	char tmpbuff[end-strt+2];
+	strcpy(tmpbuff,expr);
+
+	strncpy(tmpbuff,expr+strt,end-strt+1);
+	tmpbuff[end-strt+1] = '\0';
+	
+	strcpy(buff,tmpbuff);
 	return buff;
 }
